server/src-lib/Hasura/RQL/DDL/Remote/Validate.hs

```diff
@@ -0,0 +1,201 @@
+validateRelationship ::
+     CreateRemoteRelationship
+  -> GC.GCtx
+  -> HashMap QualifiedTable TableInfo
+  -> Either (NonEmpty ValidationError) ()
+validateRelationship createRemoteRelationship gctx tables = do
+  case HM.lookup tableName tables of
+    Nothing -> Left (pure (TableNotFound tableName))
+    Just table -> do
+      fieldInfos <-
+        fmap
+          HM.fromList
+          (traverse
+             (\fieldName ->
+                case HM.lookup fieldName (tiFieldInfoMap table) of
+                  Nothing ->
+                    Left (pure (TableFieldNonexistent tableName fieldName))
+                  Just fieldInfo -> pure (fieldName, fieldInfo))
+             (toList
+                (ccrHasuraFields createRemoteRelationship)))
+      objFldInfo <-
+        lookupField
+          (ccrRemoteField createRemoteRelationship)
+          (GS._gQueryRoot gctx)
+      case VT._fiLoc objFldInfo of
+        HasuraType ->
+          Left
+            (pure
+               (FieldNotFoundInRemoteSchema
+                  (ccrRemoteField createRemoteRelationship)))
+        RemoteType {} ->
+          toEither
+            (validateRemoteArguments
+               tables
+               (VT._fiParams objFldInfo)
+               (remoteArgumentsToMap
+                  (ccrRemoteArguments
+                     createRemoteRelationship))
+               (HM.fromList
+                  (map (first fieldNameToVariable) (HM.toList fieldInfos))))
+  where
+    tableName = ccrTable createRemoteRelationship
+
+-- | Convert a field name to a variable name.
+fieldNameToVariable :: FieldName -> G.Variable
+fieldNameToVariable = G.Variable . G.Name . getFieldNameTxt
+
+-- | Lookup the field in the schema.
+lookupField ::
+     G.Name
+  -> VT.ObjTyInfo
+  -> Either (NonEmpty ValidationError) VT.ObjFldInfo
+lookupField name objFldInfo = viaObject objFldInfo
+  where
+    viaObject =
+      maybe (Left (pure (CouldntFindRemoteField name objFldInfo))) pure .
+      HM.lookup name .
+      VT._otiFields
+
+-- | Validate remote input arguments against the remote schema.
+validateRemoteArguments ::
+     HashMap QualifiedTable TableInfo
+  -> HashMap G.Name InpValInfo
+  -> HashMap G.Name G.Value
+  -> HashMap G.Variable FieldInfo
+  -> Validation (NonEmpty ValidationError) ()
+validateRemoteArguments tables expectedArguments providedArguments permittedVariables = do
+  traverse validateProvided (HM.toList providedArguments)
+  traverse validateExpected (HM.toList expectedArguments)
+  pure ()
+  where
+    validateProvided (providedKey, providedValue) =
+      case HM.lookup providedKey expectedArguments of
+        Nothing -> Failure (pure (NoSuchArgumentForRemote providedKey))
+        Just expectedType ->
+          validateType
+            tables
+            permittedVariables
+            providedValue
+            (_iviType expectedType)
+    validateExpected (expectedKey, expectedInpValInfo) =
```

Comment:
we need to check this only for arguments that are required. 

Check `isNullable :: GType -> Bool` to check whether a GType is nullable.

Reply:
Great, will do.